// mind-lake-sdk/dist/types/index.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/types/index.ts
var types_exports = {};
__export(types_exports, {
  DataType: () => DataType
});
module.exports = __toCommonJS(types_exports);
var DataType = /* @__PURE__ */ ((DataType2) => {
  DataType2[DataType2["int4"] = 1] = "int4";
  DataType2[DataType2["int8"] = 2] = "int8";
  DataType2[DataType2["float4"] = 3] = "float4";
  DataType2[DataType2["float8"] = 4] = "float8";
  DataType2[DataType2["decimal"] = 5] = "decimal";
  DataType2[DataType2["text"] = 6] = "text";
  DataType2[DataType2["timestamp"] = 7] = "timestamp";
  return DataType2;
})(DataType || {});
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  DataType
});
