// everpay/esm/index.js
import { getEverpayTxMessage, signMessageAsync, transferAsync } from './lib/sign';
import { getSwapInfo, getEverpayBalance, getEverpayBalances, getEverpayInfo, getEverpayTransaction, getEverpayTransactions, getExpressInfo, getMintdEverpayTransactionByChainTxHash, postTx, getSwapPrice, placeSwapOrder, getFees, getFee } from './api';
import { everpayTxVersion, getExpressHost, getEverpayHost, getSwapHost } from './config';
import { getTimestamp, getTokenBySymbol, toBN, getAccountChainType, fromDecimalToUnit, genTokenTag, matchTokenTag, genExpressData, fromUnitToDecimalBN, genBundleData, getTokenBurnFeeByChainType, getChainDecimalByChainType, isArweaveChainPSTMode } from './utils/util';
import { checkParams } from './utils/check';
import { ERRORS } from './utils/errors';
import { utils } from 'ethers';
import { EverpayBase, EverpayAction, ChainType } from './types';
import { swapParamsClientToServer, swapParamsServerToClient } from './utils/swap';
export * from './types';
class Everpay extends EverpayBase {
    constructor(config) {
        var _a, _b;
        super();
        this.getAccountChainType = getAccountChainType;
        this.cacheHelper = async (key) => {
            var _a, _b, _c;
            const timestamp = getTimestamp();
            // cache info 3 mins
            if (((_a = this._cachedInfo[key]) === null || _a === void 0 ? void 0 : _a.value) != null &&
                this._cachedInfo[key].timestamp > timestamp - 3 * 60) {
                return (_b = this._cachedInfo[key]) === null || _b === void 0 ? void 0 : _b.value;
            }
            if (key === 'everpay') {
                const value = await await getEverpayInfo(this._apiHost);
                this._cachedInfo[key] = { value, timestamp };
            }
            else if (key === 'express') {
                const value = await await getExpressInfo(this._expressHost);
                this._cachedInfo[key] = { value, timestamp };
            }
            else if (key === 'swap') {
                const value = await await getSwapInfo(this._swapHost);
                this._cachedInfo[key] = { value, timestamp };
            }
            return (_c = this._cachedInfo[key]) === null || _c === void 0 ? void 0 : _c.value;
        };
        this._config = {
            ...config,
            account: (_a = config === null || config === void 0 ? void 0 : config.account) !== null && _a !== void 0 ? _a : '',
            chainType: (_b = config === null || config === void 0 ? void 0 : config.chainType) !== null && _b !== void 0 ? _b : ChainType.ethereum
        };
        this._apiHost = getEverpayHost(config === null || config === void 0 ? void 0 : config.debug);
        this._expressHost = getExpressHost(config === null || config === void 0 ? void 0 : config.debug);
        this._swapHost = getSwapHost(config === null || config === void 0 ? void 0 : config.debug);
        this._cachedInfo = {};
    }
    async info() {
        const result = await this.cacheHelper('everpay');
        return result;
    }
    async expressInfo() {
        const result = await this.cacheHelper('express');
        return result;
    }
    async swapInfo() {
        const result = await this.cacheHelper('swap');
        return result;
    }
    async balance(params) {
        var _a, _b;
        await this.info();
        const { symbol, account } = params;
        const acc = account !== null && account !== void 0 ? account : this._config.account;
        const token = getTokenBySymbol(symbol, (_b = (_a = this._cachedInfo) === null || _a === void 0 ? void 0 : _a.everpay) === null || _b === void 0 ? void 0 : _b.value.tokenList);
        checkParams({ account: acc, symbol, token });
        const mergedParams = {
            tokenTag: genTokenTag(token),
            account: acc
        };
        const everpayBalance = await getEverpayBalance(this._apiHost, mergedParams);
        return fromDecimalToUnit(everpayBalance.balance.amount, everpayBalance.balance.decimals);
    }
    async balances(params) {
        const info = await this.info();
        params = (params !== null && params !== void 0 ? params : {});
        const { account } = params;
        const acc = account !== null && account !== void 0 ? account : this._config.account;
        checkParams({ account: acc });
        const mergedParams = {
            account: acc
        };
        const everpayBalances = await getEverpayBalances(this._apiHost, mergedParams);
        const balances = everpayBalances.balances.map(item => {
            const tag = item.tag;
            const token = info.tokenList.find(token => token.tag === tag);
            return {
                chainType: token === null || token === void 0 ? void 0 : token.chainType,
                symbol: token === null || token === void 0 ? void 0 : token.symbol.toUpperCase(),
                address: token.id,
                balance: fromDecimalToUnit(item.amount, item.decimals)
            };
        });
        return balances;
    }
    async getMergedTxsParams(params) {
        var _a, _b;
        const { page, symbol, action, withoutAction } = params;
        const mergedParams = {};
        if (page !== undefined) {
            mergedParams.page = page;
        }
        if (symbol !== undefined) {
            await this.info();
            const token = getTokenBySymbol(symbol, (_b = (_a = this._cachedInfo) === null || _a === void 0 ? void 0 : _a.everpay) === null || _b === void 0 ? void 0 : _b.value.tokenList);
            checkParams({ token });
            mergedParams.tokenId = token.id;
        }
        if (action !== undefined) {
            checkParams({ action });
            mergedParams.action = action;
        }
        if (withoutAction !== undefined) {
            mergedParams.withoutAction = withoutAction;
        }
        return mergedParams;
    }
    async txs(params) {
        const mergedParams = await this.getMergedTxsParams(params);
        return await getEverpayTransactions(this._apiHost, mergedParams);
    }
    async txsByAccount(params) {
        var _a, _b;
        checkParams({ account: (_a = params.account) !== null && _a !== void 0 ? _a : this._config.account });
        const mergedParams = await this.getMergedTxsParams(params);
        mergedParams.account = (_b = params.account) !== null && _b !== void 0 ? _b : this._config.account;
        return await getEverpayTransactions(this._apiHost, mergedParams);
    }
    async txByHash(everHash) {
        checkParams({ everHash });
        return await getEverpayTransaction(this._apiHost, everHash);
    }
    async mintedTxByChainTxHash(chainTxHash) {
        checkParams({ chainTxHash });
        return await getMintdEverpayTransactionByChainTxHash(this._apiHost, chainTxHash);
    }
    async fees() {
        return await getFees(this._apiHost);
    }
    async fee(symbol) {
        var _a, _b;
        await this.info();
        const token = getTokenBySymbol(symbol, (_b = (_a = this._cachedInfo) === null || _a === void 0 ? void 0 : _a.everpay) === null || _b === void 0 ? void 0 : _b.value.tokenList);
        checkParams({ symbol, token });
        return await getFee(this._apiHost, genTokenTag(token));
    }
    async deposit(params) {
        var _a, _b, _c;
        await this.info();
        const { amount, symbol } = params;
        const from = this._config.account;
        const token = getTokenBySymbol(symbol, (_b = (_a = this._cachedInfo) === null || _a === void 0 ? void 0 : _a.everpay) === null || _b === void 0 ? void 0 : _b.value.tokenList);
        const chainType = this._config.chainType;
        checkParams({ account: from, symbol, token, amount });
        // arweave 上的 PST 充值必须是整数
        if (isArweaveChainPSTMode(token) && chainType === ChainType.arweave && parseInt(amount) !== +amount) {
            throw new Error(ERRORS.DEPOSIT_ARWEAVE_PST_MUST_BE_INTEGER);
        }
        const chainDecimal = getChainDecimalByChainType(token, chainType);
        const value = utils.parseUnits(toBN(amount).toString(), chainDecimal);
        return await transferAsync(this._config, (_c = this._cachedInfo.everpay) === null || _c === void 0 ? void 0 : _c.value, {
            symbol,
            token,
            from: from !== null && from !== void 0 ? from : '',
            value
        });
    }
    // amount 为实际收款数量
    async getEverpayTxWithoutSig(type, params) {
        var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p;
        await this.info();
        const { symbol, amount, fee, quickMode } = params;
        const token = getTokenBySymbol(symbol, (_b = (_a = this._cachedInfo) === null || _a === void 0 ? void 0 : _a.everpay) === null || _b === void 0 ? void 0 : _b.value.tokenList);
        const from = this._config.account;
        let data = params.data;
        let to = params === null || params === void 0 ? void 0 : params.to;
        let decimalFeeBN = toBN(0);
        let decimalOperateAmountBN = toBN(0);
        let action = EverpayAction.transfer;
        checkParams({ account: from, symbol, token, to });
        if (type === 'transfer') {
            checkParams({ amount });
            action = EverpayAction.transfer;
            decimalOperateAmountBN = fromUnitToDecimalBN(amount, (_c = token === null || token === void 0 ? void 0 : token.decimals) !== null && _c !== void 0 ? _c : 0);
        }
        else if (type === 'bundle') {
            action = EverpayAction.bundle;
            decimalOperateAmountBN = fromUnitToDecimalBN(amount, (_d = token === null || token === void 0 ? void 0 : token.decimals) !== null && _d !== void 0 ? _d : 0);
            // eslint-disable-next-line @typescript-eslint/strict-boolean-expressions
        }
        else if (type === 'withdraw') {
            checkParams({ amount });
            const chainType = params.chainType;
            // PST 提现到 arweave 网络必须是整数
            if (isArweaveChainPSTMode(token) && chainType === ChainType.arweave && parseInt(amount) !== +amount) {
                throw new Error(ERRORS.PST_WITHDARW_TO_ARWEAVE_MUST_BE_INTEGER);
            }
            const balance = await this.balance({ symbol });
            const decimalBalanceBN = fromUnitToDecimalBN(balance, (_e = token === null || token === void 0 ? void 0 : token.decimals) !== null && _e !== void 0 ? _e : 0);
            // 快速提现
            if (quickMode === true) {
                action = EverpayAction.transfer;
                const expressInfo = await this.expressInfo();
                const tokenTag = genTokenTag(token);
                const foundExpressTokenData = expressInfo.tokens.find(t => matchTokenTag(tokenTag, t.tokenTag));
                if (foundExpressTokenData == null) {
                    throw new Error(ERRORS.WITHDRAW_TOKEN_NOT_SUPPORT_QUICK_MODE);
                }
                const quickWithdrawLimitBN = fromUnitToDecimalBN(foundExpressTokenData.walletBalance, (_f = token === null || token === void 0 ? void 0 : token.decimals) !== null && _f !== void 0 ? _f : 0);
                // 快速提现的手续费，只放入 data 字段中
                const quickWithdrawFeeBN = fee !== undefined
                    ? fromUnitToDecimalBN(fee, (_g = token === null || token === void 0 ? void 0 : token.decimals) !== null && _g !== void 0 ? _g : 0)
                    : toBN(foundExpressTokenData.withdrawFee);
                // 快速提现的 amount 为全部数量
                decimalOperateAmountBN = fromUnitToDecimalBN(amount, (_h = token === null || token === void 0 ? void 0 : token.decimals) !== null && _h !== void 0 ? _h : 0).plus(quickWithdrawFeeBN);
                if (decimalOperateAmountBN.gt(decimalBalanceBN)) {
                    throw new Error(ERRORS.WITHDRAW_AMOUNT_LESS_THAN_FEE);
                }
                if (decimalOperateAmountBN.gt(quickWithdrawLimitBN)) {
                    throw new Error(ERRORS.INSUFFICIENT_QUICK_WITHDRAWAL_AMOUNT);
                }
                const expressData = genExpressData({
                    chainType, to, fee: quickWithdrawFeeBN.toString()
                });
                data = data !== undefined ? { ...data, ...expressData } : { ...expressData };
                // to 需要更改为快速提现收款账户
                to = expressInfo.address;
                // 普通提现
            }
            else {
                action = EverpayAction.withdraw;
                if (fee !== undefined) {
                    decimalFeeBN = fromUnitToDecimalBN(fee, (_j = token === null || token === void 0 ? void 0 : token.decimals) !== null && _j !== void 0 ? _j : 0);
                }
                else {
                    const feeItem = await getFee(this._apiHost, genTokenTag(token));
                    decimalFeeBN = toBN((_k = getTokenBurnFeeByChainType(token, feeItem, chainType)) !== null && _k !== void 0 ? _k : '0');
                }
                const targetChainType = chainType;
                data = data !== undefined ? { ...data, targetChainType } : { targetChainType };
                decimalOperateAmountBN = fromUnitToDecimalBN(amount, (_l = token === null || token === void 0 ? void 0 : token.decimals) !== null && _l !== void 0 ? _l : 0);
                if (decimalOperateAmountBN.plus(decimalFeeBN).gt(decimalBalanceBN)) {
                    throw new Error(ERRORS.WITHDRAW_AMOUNT_LESS_THAN_FEE);
                }
            }
        }
        const everpayTxWithoutSig = {
            tokenSymbol: symbol,
            action,
            from,
            to,
            amount: decimalOperateAmountBN.toString(),
            fee: decimalFeeBN.toString(),
            feeRecipient: (_p = (_o = (_m = this._cachedInfo) === null || _m === void 0 ? void 0 : _m.everpay) === null || _o === void 0 ? void 0 : _o.value.feeRecipient) !== null && _p !== void 0 ? _p : '',
            nonce: Date.now().toString(),
            tokenID: token === null || token === void 0 ? void 0 : token.id,
            chainType: token === null || token === void 0 ? void 0 : token.chainType,
            chainID: token === null || token === void 0 ? void 0 : token.chainID,
            data: data !== undefined ? JSON.stringify(data) : '',
            version: everpayTxVersion
        };
        return everpayTxWithoutSig;
    }
    getEverpayTxMessage(everpayTxWithoutSig) {
        return getEverpayTxMessage(everpayTxWithoutSig);
    }
    async signedEverpayTx(everpayTxWithoutSig) {
        const messageData = getEverpayTxMessage(everpayTxWithoutSig);
        const { sig, everHash } = await signMessageAsync(this._config, messageData);
        const everpayTx = {
            ...everpayTxWithoutSig,
            sig
        };
        return { everpayTx, everHash };
    }
    async sendEverpayTx(everpayTxWithoutSig) {
        const { everpayTx, everHash } = await this.signedEverpayTx(everpayTxWithoutSig);
        const postEverpayTxResult = await postTx(this._apiHost, everpayTx);
        return {
            ...postEverpayTxResult,
            everpayTx,
            everHash
        };
    }
    async transfer(params) {
        const everpayTxWithoutSig = await this.getEverpayTxWithoutSig('transfer', params);
        return await this.sendEverpayTx(everpayTxWithoutSig);
    }
    async withdraw(params) {
        var _a;
        await this.info();
        const to = (_a = params.to) !== null && _a !== void 0 ? _a : this._config.account;
        const everpayTxWithoutSig = await this.getEverpayTxWithoutSig('withdraw', {
            ...params,
            to
        });
        return await this.sendEverpayTx(everpayTxWithoutSig);
    }
    async getBundleData(items, expiration) {
        var _a, _b;
        await this.info();
        return genBundleData({
            items,
            tokenList: (_b = (_a = this._cachedInfo.everpay) === null || _a === void 0 ? void 0 : _a.value) === null || _b === void 0 ? void 0 : _b.tokenList,
            // 设置 60s 过期
            expiration: expiration !== null && expiration !== void 0 ? expiration : Math.round(Date.now() / 1000) + 60
        });
    }
    async signBundleData(bundleData) {
        const { items, expiration, salt, version } = bundleData;
        const { sig } = await signMessageAsync(this._config, JSON.stringify({
            // 只签名这几个字段，并且顺序需要保持一致
            items, expiration, salt, version
        }));
        const sigs = bundleData.sigs != null ? bundleData.sigs : {};
        sigs[this._config.account] = sig;
        return {
            items, expiration, salt, version, sigs
        };
    }
    async bundle(params) {
        const everpayTxWithoutSig = await this.getEverpayTxWithoutSig('bundle', params);
        return await this.sendEverpayTx(everpayTxWithoutSig);
    }
    async swapPrice(params) {
        var _a, _b;
        await Promise.all([this.info(), this.swapInfo()]);
        const everpayInfo = (_a = this._cachedInfo.everpay) === null || _a === void 0 ? void 0 : _a.value;
        const swapInfo = (_b = this._cachedInfo.swap) === null || _b === void 0 ? void 0 : _b.value;
        const paramsToServer = swapParamsClientToServer(params, everpayInfo, swapInfo);
        const result = await getSwapPrice(this._swapHost, paramsToServer);
        const formated = swapParamsServerToClient(result, everpayInfo, swapInfo);
        return {
            ...formated,
            indicativePrice: result.indicativePrice,
            spreadPercent: result.spreadPercent
        };
    }
    async getSwapBundleData(params) {
        var _a;
        await Promise.all([this.info(), this.swapInfo()]);
        const swapInfo = (_a = this._cachedInfo.swap) === null || _a === void 0 ? void 0 : _a.value;
        const { tokenIn, tokenOut, tokenInAmount, tokenOutAmount } = params;
        return await this.getBundleData([
            {
                from: this._config.account,
                to: swapInfo.address,
                symbol: tokenIn,
                amount: tokenInAmount
            },
            {
                from: swapInfo.address,
                to: this._config.account,
                symbol: tokenOut,
                amount: tokenOutAmount
            }
        ]);
    }
    async swapOrder(bundleData) {
        const { items, expiration, salt, version } = bundleData;
        const { sig } = await signMessageAsync(this._config, JSON.stringify(bundleData));
        return await placeSwapOrder(this._swapHost, {
            items,
            expiration,
            salt,
            version,
            sigs: {
                [this._config.account]: sig
            }
        });
    }
}
export default Everpay;
