// arseeding-js/esm/manifest.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.checkPaths = exports.walkpath = exports.generateManifest = void 0;
const fs_1 = require("fs");
const path_1 = __importDefault(require("path"));
const generateManifest = (config) => {
    const { items, indexFile } = config;
    const manifest = {
        manifest: 'arweave/paths',
        version: '0.1.0',
        paths: {}
    };
    if (indexFile && items.has(indexFile)) {
        manifest.index = { path: indexFile };
    }
    else if (items.has('index.html')) {
        manifest.index = { path: 'index.html' };
    }
    for (const [k, v] of items.entries()) {
        manifest.paths[k] = { id: v };
    }
    return manifest;
};
exports.generateManifest = generateManifest;
async function* walkpath(dir) {
    for await (const d of await fs_1.promises.opendir(dir)) {
        const entry = path_1.default.join(dir, d.name);
        if (d.isDirectory())
            yield* await walkpath(entry);
        else if (d.isFile())
            yield entry;
    }
}
exports.walkpath = walkpath;
async function checkPaths(path) {
    const files = [];
    for await (const f of walkpath(path)) {
        files.push(f);
    }
    return files;
}
exports.checkPaths = checkPaths;
