// arseeding-js/esm/submitOrder.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createAndSubmitItem = void 0;
const arseeding_arbundles_1 = require("arseeding-arbundles");
const axios_1 = __importDefault(require("axios"));
async function createAndSubmitItem(data, opts, cfg) {
    const dataItem = await createAndSignItem(cfg.signer, data, opts);
    return await submit(cfg.arseedUrl, dataItem, cfg.currency, cfg.apiKey, cfg.needSeq);
}
exports.createAndSubmitItem = createAndSubmitItem;
async function createAndSignItem(signer, data, opts) {
    const dataItem = (0, arseeding_arbundles_1.createData)(data, signer, opts);
    await dataItem.sign(signer);
    return dataItem;
}
async function submit(arseedingUrl, dataItem, tokenSymbol, apiKey, needSeq) {
    const api = axios_1.default.create({ baseURL: arseedingUrl });
    let header = {
        'Content-Type': 'application/octet-stream'
    };
    if (apiKey != null) {
        header['X-API-KEY'] = apiKey;
    }
    if (needSeq) {
        header['Sort'] = 'true';
    }
    const res = await api.post(`/bundle/tx/${tokenSymbol}`, dataItem.getRaw(), {
        headers: header,
        maxBodyLength: Infinity
    });
    return res.data;
}
