// arseeding-arbundles/src/signing/keys/curve25519.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const base64url_1 = __importDefault(require("base64url"));
const ed25519_1 = require("@noble/ed25519");
const constants_1 = require("../../constants");
class Curve25519 {
    constructor(_key, pk) {
        this._key = _key;
        this.pk = pk;
        this.ownerLength = constants_1.SIG_CONFIG[2].pubLength;
        this.signatureLength = constants_1.SIG_CONFIG[2].sigLength;
        this.signatureType = 2;
    }
    get publicKey() {
        return this._publicKey;
    }
    get key() {
        return new Uint8Array(0);
    }
    sign(message) {
        return ed25519_1.sign(Buffer.from(message), Buffer.from(this.key));
    }
    static async verify(pk, message, signature) {
        let p = pk;
        if (typeof pk === "string")
            p = base64url_1.default.toBuffer(pk);
        return ed25519_1.verify(Buffer.from(signature), Buffer.from(message), Buffer.from(p));
    }
}
exports.default = Curve25519;
