// mind-lake-sdk/dist/util/abi.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/util/abi.ts
var abi_exports = {};
__export(abi_exports, {
  default: () => abi_default
});
module.exports = __toCommonJS(abi_exports);
var Abi = [
  {
    anonymous: false,
    inputs: [
      {
        indexed: true,
        internalType: "address",
        name: "wallet",
        type: "address"
      },
      {
        indexed: false,
        internalType: "bytes",
        name: "MK",
        type: "bytes"
      },
      {
        indexed: false,
        internalType: "bytes",
        name: "SK",
        type: "bytes"
      }
    ],
    name: "KeysUpdated",
    type: "event"
  },
  {
    inputs: [
      {
        internalType: "bytes",
        name: "_mk",
        type: "bytes"
      },
      {
        internalType: "bytes",
        name: "_sk",
        type: "bytes"
      }
    ],
    name: "setKeys",
    outputs: [],
    stateMutability: "nonpayable",
    type: "function"
  },
  {
    inputs: [
      {
        internalType: "address",
        name: "_wallet",
        type: "address"
      }
    ],
    name: "getKeys",
    outputs: [
      {
        internalType: "bytes",
        name: "MK",
        type: "bytes"
      },
      {
        internalType: "bytes",
        name: "SK",
        type: "bytes"
      }
    ],
    stateMutability: "view",
    type: "function"
  }
];
var abi_default = Abi;
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {});
